### PR TITLE
Add support for external names to Kubernetes Controller

### LINKED ISSUE #2230: Kubernetes Controller Support External names
### What should we add or change to make your life better?

The Kubernetes Controller supports routing currently to Services within the cluster. By using Endpoint Slices this can be made to work with Services hosted externally to the cluster if there is a list of known Ip's. However, in my case I have a DNS Name that say points to a CDN for example Azure Front Door. As such, Support for [External Name](https://github.com/HappyCodeSloth/reverse-proxy.git) type services would be best. 

#### Example

``` yaml 
apiVersion: v1
kind: Service
metadata:
  name: my-external-service
spec:
  type: ExternalName
  externalName: my.external-service.example.com
--- 
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: minimal-ingress
  annotations:
    yarp.ingress.kubernetes.io/backend-protocol: https
spec:
  ingressClassName: my-yarp-controller-ingress-class
  rules:
  - http:
      paths:
      - path: /testpath
        pathType: Prefix
        backend:
          service:
            name: my-external-service
            port:
              number: 443
```

### Why is this important to you?

Enables the Yarp Kubernetes Controller to be used to proxy to services hosted externally from the cluster without knowledge of underlying IPs using native Kubernetes Configuration.

### PR DESCRIPTION
**Pull request by HappyCodeSloth**
Closes #2230

### CODE REVIEW

**Comment by HappyCodeSloth** [comment]
[Issue](https://github.com/microsoft/reverse-proxy/issues/2230)

**Comment by HappyCodeSloth** [comment]
@microsoft-github-policy-service agree company="AVEVA"

File: src/Kubernetes.Controller/Converters/YarpParser.cs [hunk 1]
```diff
@@ -42,15 +43,75 @@ private static void HandleIngressRule(YarpIngressContext ingressContext, List<En
...
+                    if (servicePort != null)
+                    {
+                        HandleIngressRulePath(ingressContext, servicePort, endpoints, defaultSubsets, rule, path, configContext);
+                    }
                 }
             }
         }
     }
 
+    private static void HandleExternalIngressRulePath(YarpIngressContext ingressContext, string externalName, V1IngressRule rule, V1HTTPIngressPath path, YarpConfigContext configContext)
```
Comment (MihaZupan) [inline src/Kubernetes.Controller/Converters/YarpParser.cs:62]: Most of the logic here is a duplicate of `HandleIngressRulePath`.

Can we extract some of it and share it?
E.g. `CreateCluster`, `CreateRoute(YarpIngressContext, ClusterTransfer)`, `SetDestination(ClusterTransfer, address, port)`?

Comment (HappyCodeSloth) [reply src/Kubernetes.Controller/Converters/YarpParser.cs:62]: @MihaZupan Thank you for the review. I have extracted the duplicated code.

File: test/Kubernetes.Tests/Yarp.Kubernetes.Tests.csproj [hunk 2]
```diff
@@ -22,6 +22,9 @@
     <None Remove="Certificates\cert.pem" />
     <None Remove="Certificates\key.der" />
     <None Remove="Certificates\key.pem" />
+    <None Remove="testassets\external-name-ingress\clusters.json" />
+    <None Remove="testassets\external-name-ingress\ingress.yaml" />
+    <None Remove="testassets\external-name-ingress\routes.json" />
```
[the remark this answers is not in the document]
Comment (HappyCodeSloth) [reply test/Kubernetes.Tests/Yarp.Kubernetes.Tests.csproj]: Done

**Pull request merged**